File: tokscale.py
#!/usr/bin/env python3
"""Print local AI CLI token usage by model."""

from __future__ import annotations

import argparse
import json
import os
import sqlite3
import sys
from datetime import datetime, timedelta, timezone
from itertools import chain
from pathlib import Path
from zoneinfo import ZoneInfo


USER_HOME = Path.home()
CODEX_DB = USER_HOME / ".codex" / "state_5.sqlite"
CODEX_DIRS = (
    USER_HOME / ".codex" / "sessions",
    USER_HOME / ".codex" / "archived_sessions",
)
CLAUDE_PROJECTS = USER_HOME / ".claude" / "projects"
PI_SESSIONS = USER_HOME / ".pi" / "agent" / "sessions"
ANTIGRAVITY_USAGE_LOG = USER_HOME / ".gemini" / "antigravity-cli" / "tokscale-usage.jsonl"
LOCAL_TZ = ZoneInfo(os.environ.get("TOKSCALE_TZ", "Europe/Oslo"))

# USD per million tokens. Claude cache writes use the 5-minute rate because
# local transcripts do not record the cache TTL.
TOKEN_FIELDS = (
    "input_tokens",
    "output_tokens",
    "cache_read_tokens",
    "cache_write_tokens",
)
CLAUDE_RATES = {
    "claude-fable-5": (10.0, 50.0, 1.0, 12.5),
    "claude-opus-5": (5.0, 25.0, 0.5, 6.25),
    "claude-sonnet-5": (2.0, 10.0, 0.2, 2.5),
}

# API-equivalent USD per million total tokens used by this report.
OPENAI_RATES = {
    "gpt-5.6-sol": 0.8,
    "codex-auto-review": 0.459224,
    "gpt-5.6-luna": 0.0177545,
}

# API-equivalent proxy rates for Antigravity's Gemini models, USD per million
# tokens. Gemini 3.6/3.7 introductory standard pricing applies through 2026.
GEMINI_PROXY_RATES = {
    "gemini-3.7-flash": (0.75, 3.75, 0.075),
    "gemini-3.6-flash": (0.75, 3.75, 0.075),
    "gemini-3.5-flash": (1.5, 9.0, 0.15),
}


def positive_int(value) -> int:
    try:
        return max(0, int(value or 0))
    except (TypeError, ValueError):
        return 0


def local_time(value) -> datetime:
    if isinstance(value, (int, float)):
        seconds = value / 1000 if value > 10_000_000_000 else value
        return datetime.fromtimestamp(seconds, tz=timezone.utc).astimezone(LOCAL_TZ)
    try:
        parsed = datetime.fromisoformat(str(value or "").replace("Z", "+00:00"))
    except ValueError:
        return datetime.now(LOCAL_TZ)
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    return parsed.astimezone(LOCAL_TZ)


def usage_event(
    model,
    timestamp,
    *,
    provider=None,
    input_tokens=0,
    output_tokens=0,
    cache_read=0,
    cache_write=0,
    total=None,
    cost_usd=None,
) -> dict:
    input_tokens = positive_int(input_tokens)
    output_tokens = positive_int(output_tokens)
    cache_read = positive_int(cache_read)
    cache_write = positive_int(cache_write)
    calculated_total = input_tokens + output_tokens + cache_read + cache_write
    return {
        "model": model or "unknown",
        "provider": provider or "unknown",
        "timestamp": local_time(timestamp),
        "input_tokens": input_tokens,
        "output_tokens": output_tokens,
        "cache_read_tokens": cache_read,
        "cache_write_tokens": cache_write,
        "tokens": positive_int(total) if total is not None else calculated_total,
        "cost_usd": float(cost_usd) if cost_usd is not None else None,
    }


def scan_codex_rollouts():
    """Read per-turn usage; cached input is part of input, not extra tokens."""
    for base in CODEX_DIRS:
        if not base.exists():
            continue
        for path in base.rglob("*.jsonl"):
            model = None
            try:
                with path.open("r", encoding="utf-8", errors="replace") as stream:
                    for line in stream:
                        try:
                            record = json.loads(line)
                        except json.JSONDecodeError:
                            continue
                        payload = record.get("payload") or {}
                        if record.get("type") == "turn_context" and payload.get("model"):
                            model = payload["model"]
                        usage = (payload.get("info") or {}).get("last_token_usage")
                        if not model or not usage:
                            continue
                        yield usage_event(
                            model,
                            record.get("timestamp"),
                            provider="codex",
                            input_tokens=usage.get("input_tokens"),
                            output_tokens=usage.get("output_tokens"),
                            cache_read=usage.get("cached_input_tokens"),
                            cache_write=usage.get("cache_write_input_tokens"),
                            total=usage.get("total_tokens"),
                        )
            except OSError:
                continue


def scan_codex():
    rollouts = iter(scan_codex_rollouts())
    first = next(rollouts, None)
    if first is not None:
        yield first
        yield from rollouts
        return
    if not CODEX_DB.exists():
        return

    database = None
    try:
        database = sqlite3.connect(f"file:{CODEX_DB}?mode=ro", uri=True)
        database.row_factory = sqlite3.Row
        for row in database.execute(
            "SELECT model, updated_at, tokens_used "
            "FROM threads WHERE tokens_used > 0"
        ):
            yield usage_event(
                row["model"], row["updated_at"], provider="codex", total=row["tokens_used"]
            )
    except sqlite3.Error as error:
        print(f"warning: could not read Codex database: {error}", file=sys.stderr)
    finally:
        if database is not None:
            database.close()


def scan_claude():
    if not CLAUDE_PROJECTS.exists():
        return

    for path in CLAUDE_PROJECTS.rglob("*.jsonl"):
        try:
            with path.open("r", encoding="utf-8", errors="replace") as stream:
                for line in stream:
                    try:
                        record = json.loads(line)
                    except json.JSONDecodeError:
                        continue
                    if record.get("type") != "assistant":
                        continue
                    message = record.get("message") or {}
                    usage = message.get("usage") or {}
                    model = message.get("model")
                    if not usage or not model or model == "<synthetic>":
                        continue
                    yield usage_event(
                        model,
                        record.get("timestamp"),
                        provider="claude",
                        input_tokens=usage.get("input_tokens"),
                        output_tokens=usage.get("output_tokens"),
                        cache_read=usage.get("cache_read_input_tokens"),
                        cache_write=usage.get("cache_creation_input_tokens"),
                    )
        except OSError:
            continue


def scan_pi():
    """Read Pi assistant responses, including provider and OpenRouter cost."""
    if not PI_SESSIONS.exists():
        return
    for path in PI_SESSIONS.rglob("*.jsonl"):
        try:
            with path.open("r", encoding="utf-8", errors="replace") as stream:
                for line in stream:
                    try:
                        record = json.loads(line)
                    except json.JSONDecodeError:
                        continue
                    message = record.get("message") or {}
                    usage = message.get("usage") or {}
                    if record.get("type") != "message" or message.get("role") != "assistant":
                        continue
                    model = message.get("model")
                    if not model or not usage:
                        continue
                    cost = usage.get("cost") or {}
                    yield usage_event(
                        model,
                        record.get("timestamp"),
                        provider=message.get("provider") or "pi",
                        input_tokens=usage.get("input"),
                        output_tokens=usage.get("output"),
                        cache_read=usage.get("cacheRead"),
                        cache_write=usage.get("cacheWrite"),
                        total=usage.get("totalTokens"),
                        cost_usd=cost.get("total"),
                    )
        except OSError:
            continue


def scan_antigravity():
    """Read opt-in status-line snapshots and emit cumulative deltas per turn."""
    if not ANTIGRAVITY_USAGE_LOG.exists():
        return
    snapshots = {}
    try:
        with ANTIGRAVITY_USAGE_LOG.open("r", encoding="utf-8", errors="replace") as stream:
            for line in stream:
                try:
                    record = json.loads(line)
                except json.JSONDecodeError:
                    continue
                payload = record.get("payload") or record
                context = payload.get("context_window") or {}
                conversation = (
                    payload.get("conversation_id")
                    or payload.get("session_id")
                    or payload.get("transcript_path")
                )
                model_info = payload.get("model") or {}
                model = model_info.get("id") or model_info.get("display_name")
                timestamp = record.get("captured_at") or payload.get("created_at")
                if not conversation or not model or not timestamp:
                    continue
                input_total = positive_int(context.get("total_input_tokens"))
                output_total = positive_int(context.get("total_output_tokens"))
                if not input_total and not output_total:
                    continue
                current = (input_total, output_total, model, timestamp)
                previous = snapshots.get(conversation)
                snapshots[conversation] = current
                if previous is None:
                    delta_input, delta_output = input_total, output_total
                else:
                    delta_input = max(0, input_total - previous[0])
                    delta_output = max(0, output_total - previous[1])
                if delta_input or delta_output:
                    yield usage_event(
                        model,
                        timestamp,
                        provider="antigravity",
                        input_tokens=delta_input,
                        output_tokens=delta_output,
                        total=delta_input + delta_output,
                    )
    except OSError:
        return


def event_cost(event: dict) -> float | None:
    if event.get("cost_usd") is not None:
        return max(0.0, float(event["cost_usd"]))
    model = event["model"]
    if event.get("provider") == "antigravity":
        normalized = model.lower().replace(" ", "-")
        for name, rates in GEMINI_PROXY_RATES.items():
            if name in normalized:
                return (
                    event["input_tokens"] * rates[0]
                    + event["output_tokens"] * rates[1]
                    + event["cache_read_tokens"] * rates[2]
                ) / 1_000_000
    if model in CLAUDE_RATES:
        return sum(
            event[field] * rate
            for field, rate in zip(TOKEN_FIELDS, CLAUDE_RATES[model])
        ) / 1_000_000
    if model in OPENAI_RATES:
        return event["tokens"] * OPENAI_RATES[model] / 1_000_000
    return None


def period_start(period: str) -> datetime | None:
    if period == "all":
        return None
    now = datetime.now(LOCAL_TZ)
    start = now.replace(hour=0, minute=0, second=0, microsecond=0)
    if period == "weekly":
        return start - timedelta(days=start.weekday())
    if period == "monthly":
        return start.replace(day=1)
    return start


def collect(period: str) -> tuple[list[dict], datetime | None, datetime | None]:
    since = period_start(period)
    models = {}
    first = last = None
    for event in chain(scan_codex(), scan_claude(), scan_pi(), scan_antigravity()):
        timestamp = event["timestamp"]
        if since and timestamp < since:
            continue
        first = timestamp if first is None or timestamp < first else first
        last = timestamp if last is None or timestamp > last else last
        key = (event["provider"], event["model"])
        row = models.setdefault(
            key,
            {
                "model": event["model"],
                "provider": event["provider"],
                "tokens": 0,
                **dict.fromkeys(TOKEN_FIELDS, 0),
                "events": 0,
                "cost": 0.0,
                "cost_known": True,
            },
        )
        for field in ("tokens", *TOKEN_FIELDS):
            row[field] += event[field]
        row["events"] += 1
        cost = event_cost(event)
        if cost is None:
            row["cost_known"] = False
        else:
            row["cost"] += cost

    ordered = sorted(models.values(), key=lambda row: row["tokens"], reverse=True)
    return ordered, first, last


def format_tokens(value: int) -> str:
    if value >= 1_000_000_000:
        return f"{value / 1_000_000_000:.1f}B"
    if value >= 1_000_000:
        return f"{value / 1_000_000:.1f}M"
    if value >= 1_000:
        return f"{value / 1_000:.1f}K"
    return str(value)


def format_cost(value: float, known: bool) -> str:
    if not known:
        return "—"
    if value >= 1000:
        return f"${value / 1000:.1f}K"
    if 0 < value < 0.01:
        return f"${value:.3f}"
    return f"${value:.2f}"


def interval(first: datetime | None, last: datetime | None) -> str:
    if not first or not last:
        return "No data"
    first_label = f"{first.strftime('%b')} {first.day}"
    last_label = f"{last.strftime('%b')} {last.day}"
    return first_label if first.date() == last.date() else f"{first_label} – {last_label}"


def print_models(breakdown: bool, period: str) -> None:
    models, first, last = collect(period)
    print(f"tokscale · Models · {period}")
    print(f"range: {interval(first, last)}")
    print()
    print(f"{'Model':<30} {'Tokens':>12} {'Cost':>12}")
    print("-" * 58)
    for model in models:
        cost = format_cost(model["cost"], model["cost_known"])
        label = model["model"]
        if model["provider"] not in ("codex", "claude"):
            label = f"{model['provider']}/{label}"
        print(f"{label:<30} {format_tokens(model['tokens']):>12} {cost:>12}")

    if not breakdown:
        return
    print()
    print("Breakdown")
    print("---------")
    print(
        f"{'Model':<24} {'Input':>10} {'Output':>10} "
        f"{'Cache read':>12} {'Cache write':>13} {'Events':>8}"
    )
    print("-" * 83)
    for model in models:
        print(
            f"{model['model']:<24} "
            f"{format_tokens(model['input_tokens']):>10} "
            f"{format_tokens(model['output_tokens']):>10} "
            f"{format_tokens(model['cache_read_tokens']):>12} "
            f"{format_tokens(model['cache_write_tokens']):>13} "
            f"{model['events']:>8}"
        )
    print()
    print("Note: Codex cache read is included inside Input and is not added twice.")


def main() -> None:
    parser = argparse.ArgumentParser(description="Report local Codex and Claude token usage")
    commands = parser.add_subparsers(dest="command", required=True)
    models = commands.add_parser("models", help="show usage grouped by model")
    models.add_argument("--breakdown", action="store_true", help="show token categories")
    periods = models.add_mutually_exclusive_group()
    for period in ("daily", "weekly", "monthly", "all"):
        periods.add_argument(f"--{period}", dest="period", action="store_const", const=period)
    args = parser.parse_args()
    print_models(args.breakdown, args.period or "all")


if __name__ == "__main__":
    try:
        main()
    except BrokenPipeError:
        pass

File: go/tokscale.go
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"io/fs"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"
	"unicode"
)

type usage struct {
	Input       int64 `json:"input_tokens"`
	Output      int64 `json:"output_tokens"`
	CacheRead   int64 `json:"cache_read_input_tokens"`
	CacheCreate int64 `json:"cache_creation_input_tokens"`
	CachedInput int64 `json:"cached_input_tokens"`
	CacheWrite  int64 `json:"cache_write_input_tokens"`
	Total       int64 `json:"total_tokens"`
}

type record struct {
	Type      string `json:"type"`
	Timestamp any    `json:"timestamp"`
	Payload   struct {
		Model string `json:"model"`
		Info  struct {
			Last *usage `json:"last_token_usage"`
		} `json:"info"`
	} `json:"payload"`
	Message struct {
		Model string `json:"model"`
		Usage *usage `json:"usage"`
	} `json:"message"`
}

type piRecord struct {
	Type      string `json:"type"`
	Timestamp any    `json:"timestamp"`
	Message   struct {
		Role     string   `json:"role"`
		Provider string   `json:"provider"`
		Model    string   `json:"model"`
		Usage    *piUsage `json:"usage"`
	} `json:"message"`
}

type piUsage struct {
	Input      int64   `json:"input"`
	Output     int64   `json:"output"`
	CacheRead  int64   `json:"cacheRead"`
	CacheWrite int64   `json:"cacheWrite"`
	Total      int64   `json:"totalTokens"`
	Cost       *piCost `json:"cost"`
}

type piCost struct {
	Total *float64 `json:"total"`
}

type antigravityRecord struct {
	CapturedAt     string `json:"captured_at"`
	ConversationID string `json:"conversation_id"`
	SessionID      string `json:"session_id"`
	TranscriptPath string `json:"transcript_path"`
	Model          struct {
		ID          string `json:"id"`
		DisplayName string `json:"display_name"`
	} `json:"model"`
	ContextWindow struct {
		Input  int64 `json:"total_input_tokens"`
		Output int64 `json:"total_output_tokens"`
	} `json:"context_window"`
}

type antigravitySnapshot struct {
	input, output int64
}

type event struct {
	provider                   string
	model                      string
	timestamp                  time.Time
	input, output, read, write int64
	total                      int64
	cost                       float64
	hasCost                    bool
}

type totals struct {
	model                      string
	input, output, read, write int64
	total, events              int64
	cost                       float64
	costKnown                  bool
}

type modelKey struct {
	provider, model string
}

type collector struct {
	period      string
	tz          *time.Location
	now         time.Time
	models      map[modelKey]*totals
	first, last time.Time
}

var claudeRates = map[string][4]float64{
	"claude-fable-5":  {10, 50, 1, 12.5},
	"claude-opus-5":   {5, 25, .5, 6.25},
	"claude-sonnet-5": {2, 10, .2, 2.5},
}

var openAIRates = map[string]float64{
	"gpt-5.6-sol":       .8,
	"codex-auto-review": .459224,
	"gpt-5.6-luna":      .0177545,
}

// API-equivalent proxy rates for Antigravity's Gemini models, USD per million
// tokens. Gemini 3.6/3.7 introductory standard pricing applies through 2026.
var geminiProxyRates = map[string][3]float64{
	"gemini-3.7-flash": {.75, 3.75, .075},
	"gemini-3.6-flash": {.75, 3.75, .075},
	"gemini-3.5-flash": {1.5, 9, .15},
}

func location() *time.Location {
	name := os.Getenv("TOKSCALE_TZ")
	if name == "" {
		name = "Europe/Oslo"
	}
	tz, err := time.LoadLocation(name)
	if err != nil {
		return time.Local
	}
	return tz
}

func parseTime(value any, tz *time.Location) time.Time {
	switch value := value.(type) {
	case string:
		if parsed, err := time.Parse(time.RFC3339Nano, value); err == nil {
			return parsed.In(tz)
		}
	case float64:
		seconds, nanos := int64(value), int64(0)
		if value > 10_000_000_000 {
			seconds, nanos = int64(value/1000), int64(value)%1000*1_000_000
		}
		return time.Unix(seconds, nanos).In(tz)
	}
	return time.Now().In(tz)
}

func slug(value string) string {
	return strings.Map(func(character rune) rune {
		if unicode.IsSpace(character) || character == '/' {
			return '-'
		}
		return unicode.ToLower(character)
	}, value)
}

func displayLabel(provider, model string) string {
	prefix := ""
	switch provider {
	case "codex", "claude":
	case "antigravity":
		prefix = "agy"
	case "openrouter":
		prefix = "or"
	default:
		prefix = slug(provider)
	}
	model = slug(model)
	if prefix == "" {
		return model
	}
	return prefix + "-" + model
}

func counterDelta(current, previous int64) int64 {
	if current < 0 {
		current = 0
	}
	if current < previous {
		return current
	}
	return current - previous
}

func (c *collector) includes(t time.Time) bool {
	if c.period == "all" {
		return true
	}
	now := c.now
	t = t.In(c.tz)
	switch c.period {
	case "daily":
		y1, m1, d1 := t.Date()
		y2, m2, d2 := now.Date()
		return y1 == y2 && m1 == m2 && d1 == d2
	case "weekly":
		y1, w1 := t.ISOWeek()
		y2, w2 := now.ISOWeek()
		return y1 == y2 && w1 == w2
	case "monthly":
		y1, m1, _ := t.Date()
		y2, m2, _ := now.Date()
		return y1 == y2 && m1 == m2
	}
	return false
}

func eventCost(e event) (float64, bool) {
	if e.hasCost {
		if e.cost < 0 {
			return 0, true
		}
		return e.cost, true
	}
	if e.provider == "antigravity" {
		model := strings.ReplaceAll(strings.ToLower(e.model), " ", "-")
		for name, rate := range geminiProxyRates {
			if strings.Contains(model, name) {
				return (float64(e.input)*rate[0] + float64(e.output)*rate[1] +
					float64(e.read)*rate[2]) / 1e6, true
			}
		}
	}
	if rate, ok := claudeRates[e.model]; ok {
		return (float64(e.input)*rate[0] + float64(e.output)*rate[1] +
			float64(e.read)*rate[2] + float64(e.write)*rate[3]) / 1e6, true
	}
	if rate, ok := openAIRates[e.model]; ok {
		return float64(e.total) * rate / 1e6, true
	}
	return 0, false
}

func (c *collector) add(e event) {
	if !c.includes(e.timestamp) {
		return
	}
	key := modelKey{provider: e.provider, model: e.model}
	row := c.models[key]
	if row == nil {
		row = &totals{model: displayLabel(e.provider, e.model), costKnown: true}
		c.models[key] = row
	}
	row.input += e.input
	row.output += e.output
	row.read += e.read
	row.write += e.write
	row.total += e.total
	row.events++
	if cost, ok := eventCost(e); ok {
		row.cost += cost
	} else {
		row.costKnown = false
	}
	if c.first.IsZero() || e.timestamp.Before(c.first) {
		c.first = e.timestamp
	}
	if c.last.IsZero() || e.timestamp.After(c.last) {
		c.last = e.timestamp
	}
}

func scanLines(path string, relevant func([]byte) bool, visit func(record)) {
	file, err := os.Open(path)
	if err != nil {
		return
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	scanner.Buffer(make([]byte, 64*1024), 16*1024*1024)
	for scanner.Scan() {
		if !relevant(scanner.Bytes()) {
			continue
		}
		var item record
		if json.Unmarshal(scanner.Bytes(), &item) == nil {
			visit(item)
		}
	}
}

func codexLine(line []byte) bool {
	return bytes.Contains(line, []byte(`"turn_context"`)) ||
		bytes.Contains(line, []byte(`"last_token_usage"`))
}

func usageLine(line []byte) bool {
	return bytes.Contains(line, []byte(`"assistant"`)) &&
		bytes.Contains(line, []byte(`"usage"`))
}

func walkJSONL(root string, visit func(string)) {
	_ = filepath.WalkDir(root, func(path string, entry fs.DirEntry, err error) error {
		if err == nil && !entry.IsDir() && strings.HasSuffix(path, ".jsonl") {
			visit(path)
		}
		return nil
	})
}

func scanCodex(home string, c *collector) int64 {
	var found int64
	for _, root := range []string{
		filepath.Join(home, ".codex", "sessions"),
		filepath.Join(home, ".codex", "archived_sessions"),
	} {
		walkJSONL(root, func(path string) {
			model := ""
			scanLines(path, codexLine, func(item record) {
				if item.Type == "turn_context" && item.Payload.Model != "" {
					model = item.Payload.Model
				}
				if model == "" || item.Payload.Info.Last == nil {
					return
				}
				u := item.Payload.Info.Last
				found++
				c.add(event{provider: "codex", model: model, timestamp: parseTime(item.Timestamp, c.tz),
					input: u.Input, output: u.Output, read: u.CachedInput, write: u.CacheWrite, total: u.Total})
			})
		})
	}
	return found
}

func scanCodexSQLite(home string, c *collector) {
	database := filepath.Join(home, ".codex", "state_5.sqlite")
	query := "SELECT model, updated_at, tokens_used FROM threads WHERE tokens_used > 0;"
	output, err := exec.Command("sqlite3", "-readonly", "-separator", "\t", database, query).Output()
	if err != nil {
		return
	}
	for _, line := range strings.Split(strings.TrimSpace(string(output)), "\n") {
		parts := strings.Split(line, "\t")
		if len(parts) != 3 {
			continue
		}
		timestamp, _ := strconv.ParseFloat(parts[1], 64)
		total, _ := strconv.ParseInt(parts[2], 10, 64)
		c.add(event{provider: "codex", model: parts[0], timestamp: parseTime(timestamp, c.tz), total: total})
	}
}

func scanClaude(home string, c *collector) {
	walkJSONL(filepath.Join(home, ".claude", "projects"), func(path string) {
		scanLines(path, usageLine, func(item record) {
			u, model := item.Message.Usage, item.Message.Model
			if item.Type != "assistant" || u == nil || model == "" || model == "<synthetic>" {
				return
			}
			c.add(event{provider: "claude", model: model, timestamp: parseTime(item.Timestamp, c.tz),
				input: u.Input, output: u.Output, read: u.CacheRead, write: u.CacheCreate,
				total: u.Input + u.Output + u.CacheRead + u.CacheCreate})
		})
	})
}

func scanPi(home string, c *collector) {
	walkJSONL(filepath.Join(home, ".pi", "agent", "sessions"), func(path string) {
		file, err := os.Open(path)
		if err != nil {
			return
		}
		defer file.Close()
		scanner := bufio.NewScanner(file)
		scanner.Buffer(make([]byte, 64*1024), 16*1024*1024)
		for scanner.Scan() {
			if !usageLine(scanner.Bytes()) {
				continue
			}
			var item piRecord
			if json.Unmarshal(scanner.Bytes(), &item) != nil || item.Type != "message" ||
				item.Message.Role != "assistant" || item.Message.Model == "" || item.Message.Usage == nil {
				continue
			}
			u := item.Message.Usage
			provider := item.Message.Provider
			if provider == "" {
				provider = "pi"
			}
			cost, hasCost := 0.0, false
			if u.Cost != nil && u.Cost.Total != nil {
				cost, hasCost = *u.Cost.Total, true
			}
			c.add(event{provider: provider, model: item.Message.Model, timestamp: parseTime(item.Timestamp, c.tz),
				input: u.Input, output: u.Output, read: u.CacheRead, write: u.CacheWrite, total: u.Total,
				cost: cost, hasCost: hasCost})
		}
	})
}

func scanAntigravity(home string, c *collector) {
	path := filepath.Join(home, ".gemini", "antigravity-cli", "tokscale-usage.jsonl")
	file, err := os.Open(path)
	if err != nil {
		return
	}
	defer file.Close()
	previous := make(map[string]antigravitySnapshot)
	scanner := bufio.NewScanner(file)
	scanner.Buffer(make([]byte, 64*1024), 16*1024*1024)
	for scanner.Scan() {
		var item antigravityRecord
		if json.Unmarshal(scanner.Bytes(), &item) != nil || item.CapturedAt == "" {
			continue
		}
		conversation := item.ConversationID
		if conversation == "" {
			conversation = item.SessionID
		}
		if conversation == "" {
			conversation = item.TranscriptPath
		}
		model := item.Model.ID
		if model == "" {
			model = item.Model.DisplayName
		}
		if conversation == "" || model == "" {
			continue
		}
		old := previous[conversation]
		input := max(item.ContextWindow.Input, 0)
		output := max(item.ContextWindow.Output, 0)
		deltaInput := counterDelta(input, old.input)
		deltaOutput := counterDelta(output, old.output)
		previous[conversation] = antigravitySnapshot{input, output}
		if deltaInput == 0 && deltaOutput == 0 {
			continue
		}
		c.add(event{provider: "antigravity", model: model, timestamp: parseTime(item.CapturedAt, c.tz),
			input: deltaInput, output: deltaOutput, total: deltaInput + deltaOutput})
	}
}

func tokenText(value int64) string {
	switch {
	case value >= 1e9:
		return fmt.Sprintf("%.1fB", float64(value)/1e9)
	case value >= 1e6:
		return fmt.Sprintf("%.1fM", float64(value)/1e6)
	case value >= 1e3:
		return fmt.Sprintf("%.1fK", float64(value)/1e3)
	default:
		return strconv.FormatInt(value, 10)
	}
}

func costText(row *totals) string {
	if !row.costKnown {
		return "—"
	}
	if row.cost >= 1000 {
		return fmt.Sprintf("$%.1fK", row.cost/1000)
	}
	if row.cost > 0 && row.cost < .01 {
		return fmt.Sprintf("$%.3f", row.cost)
	}
	return fmt.Sprintf("$%.2f", row.cost)
}

func interval(first, last time.Time) string {
	if first.IsZero() || last.IsZero() {
		return "no data"
	}
	start, end := fmt.Sprintf("%s %d", first.Format("Jan"), first.Day()),
		fmt.Sprintf("%s %d", last.Format("Jan"), last.Day())
	if first.Format("2006-01-02") == last.Format("2006-01-02") {
		return start
	}
	return start + " – " + end
}

func report(c *collector, breakdown bool) {
	rows := make([]*totals, 0, len(c.models))
	for _, row := range c.models {
		rows = append(rows, row)
	}
	sort.Slice(rows, func(i, j int) bool { return rows[i].total > rows[j].total })
	modelWidth := len("model")
	for _, row := range rows {
		if width := len([]rune(row.model)); width > modelWidth {
			modelWidth = width
		}
	}
	fmt.Printf("tokscale · models · %s\nrange: %s\n\n", c.period, interval(c.first, c.last))
	fmt.Printf("%-*s %12s %12s\n%s\n", modelWidth, "model", "tokens", "cost",
		strings.Repeat("-", modelWidth+26))
	for _, row := range rows {
		fmt.Printf("%-*s %12s %12s\n", modelWidth, row.model, tokenText(row.total), costText(row))
	}
	if !breakdown {
		return
	}
	fmt.Printf("\nbreakdown\n---------\n%-*s %10s %10s %12s %13s %8s\n%s\n",
		modelWidth, "model", "input", "output", "cache read", "cache write", "events",
		strings.Repeat("-", modelWidth+58))
	for _, row := range rows {
		fmt.Printf("%-*s %10s %10s %12s %13s %8d\n", modelWidth, row.model, tokenText(row.input),
			tokenText(row.output), tokenText(row.read), tokenText(row.write), row.events)
	}
	fmt.Println("\nnote: codex cache read is included inside input and is not added twice.")
}

func arguments() (bool, string) {
	args := os.Args[1:]
	if len(args) == 0 || args[0] != "models" {
		fmt.Fprintln(os.Stderr, "usage: tokscale-go models [--breakdown] [--daily|--weekly|--monthly|--all]")
		os.Exit(2)
	}
	breakdown, period := false, "all"
	for _, arg := range args[1:] {
		switch arg {
		case "--breakdown":
			breakdown = true
		case "--daily", "--weekly", "--monthly", "--all":
			period = strings.TrimPrefix(arg, "--")
		default:
			fmt.Fprintf(os.Stderr, "unknown argument: %s\n", arg)
			os.Exit(2)
		}
	}
	return breakdown, period
}

func main() {
	breakdown, period := arguments()
	home, err := os.UserHomeDir()
	if err != nil {
		panic(err)
	}
	tz := location()
	c := &collector{period: period, tz: tz, now: time.Now().In(tz), models: make(map[modelKey]*totals)}
	if scanCodex(home, c) == 0 {
		scanCodexSQLite(home, c)
	}
	scanClaude(home, c)
	scanPi(home, c)
	scanAntigravity(home, c)
	report(c, breakdown)
}
